modyn/trainer_server/internal/trainer/pytorch_trainer.py

```diff
@@ -261,45 +300,24 @@ def train(self) -> None:  # pylint: disable=too-many-locals, too-many-branches
         self._info("Handled OnBegin Callbacks.")
 
         batch_number = 0
-        for _ in range(self.epochs_per_trigger):
+        for epoch in range(self.epochs_per_trigger):
+            if self.sample_then_batch_this_epoch(epoch):
+                self.update_queue(AvailableQueues.TRAINING, batch_number, self._num_samples, training_active=False)
+                self.downsample_trigger_training_set()
+
             for batch_number, batch in enumerate(self._train_dataloader):
+                retrieve_weights_from_dataloader, weighted_optimization = self.weights_handling(len(batch))
+
                 for _, callback in self._callbacks.items():
                     callback.on_batch_begin(self._model.model, self._optimizers, batch, batch_number)
 
-                # As empty() is unreliable
-                # we try to fetch an element within 10ms. If there is no
-                # element within that timeframe returned, we continue.
-                try:
-                    req = self._status_query_queue.get(timeout=0.01)
-                    if req == TrainerMessages.STATUS_QUERY_MESSAGE:
-                        self.send_status_to_server(batch_number)
-                    elif req == TrainerMessages.MODEL_STATE_QUERY_MESSAGE:
-                        self.send_model_state_to_server()
-                    else:
-                        raise ValueError("Unknown message in the status query queue")
-                except queue.Empty:
-                    pass
-
-                sample_ids = batch[0]
-                if isinstance(sample_ids, torch.Tensor):
-                    sample_ids = sample_ids.tolist()
-                elif isinstance(sample_ids, tuple):
-                    sample_ids = list(sample_ids)
+                self.update_queue(AvailableQueues.TRAINING, batch_number, self._num_samples, training_active=True)
 
-                assert isinstance(sample_ids, list), "Cannot parse result from DataLoader"
-
-                if self._label_tranformer_function is None:
-                    target = batch[2].to(self._device)
-                else:
-                    target = self._label_tranformer_function(batch[2]).to(self._device)
+                sample_ids, target, data = self.preprocess_batch(batch)
 
-                data: Union[torch.Tensor, dict]
-                if isinstance(batch[1], torch.Tensor):
-                    data = batch[1].to(self._device)
-                elif isinstance(batch[1], dict):
-                    data: dict[str, torch.Tensor] = {}  # type: ignore[no-redef]
-                    for name, tensor in batch[1].items():
-                        data[name] = tensor.to(self._device)
+                if retrieve_weights_from_dataloader:
+                    weights = batch[3]
+                    weights = weights.float()
```

Comment:
```suggestion
                    weights = batch[3].float()
```

Reply:
also, as mentioned in my other comment, please add a very short comment why .float() is necessary